src/server/scripts/Northrend/ChamberOfAspects/RubySanctum/boss_halion.cpp

```diff
@@ -388,18 +303,44 @@ class boss_halion : public CreatureScript
                 }
             }
 
+            void SpellHit(Unit* /*who*/, SpellInfo const* spellInfo) override
+            {
+                if (spellInfo->Id == SPELL_TWILIGHT_MENDING)
+                    Talk(SAY_REGENERATE);
+            }
+
             void UpdateAI(uint32 diff) override
             {
                 if (events.IsInPhase(PHASE_TWO))
                     return;
 
-                generic_halionAI::UpdateAI(diff);
+                if (!UpdateVictim() || me->HasUnitState(UNIT_STATE_CASTING))
+                    return;
+
+                events.Update(diff);
+
+                while (uint32 eventId = events.ExecuteEvent())
+                    ExecuteEvent(eventId);
+
+                DoMeleeAttackIfReady();
             }
 
             void ExecuteEvent(uint32 eventId) override
```

Comment:
BossAI::UpdateAI  update events even if it has UNIT_STATE_CASTING, i'm having a hard time trying to understand how the cooldowns need to be handled, but that still may be needed